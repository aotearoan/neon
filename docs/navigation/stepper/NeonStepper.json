{
  "displayName": "NeonStepper",
  "description": "A stepper component for assisting users in navigating through complex flows.",
  "tags": {},
  "exportName": "default",
  "props": [
    {
      "name": "steps",
      "description": "An ordered list of the step names, these values are used for display purposes.",
      "type": {
        "name": "Array as () => Array<string>"
      },
      "required": true
    },
    {
      "name": "completedIndex",
      "description": "The index of the highest completed step (regardless of the current step the user has selected).\nThis is used to display the completed steps to the user.",
      "type": {
        "name": "number"
      },
      "defaultValue": {
        "func": false,
        "value": "-1"
      }
    },
    {
      "name": "modelValue",
      "description": "The index of the step that is currently selected.",
      "type": {
        "name": "number"
      },
      "defaultValue": {
        "func": false,
        "value": "0"
      }
    },
    {
      "name": "color",
      "description": "The color of the Stepper component.",
      "type": {
        "name": "String as () => NeonFunctionalColor"
      },
      "defaultValue": {
        "func": false,
        "value": "NeonFunctionalColor.Brand"
      }
    }
  ],
  "events": [
    {
      "name": "update:modelValue",
      "description": "Emitted when the selected step is changed.",
      "type": {
        "names": [
          "number"
        ]
      }
    }
  ]
}