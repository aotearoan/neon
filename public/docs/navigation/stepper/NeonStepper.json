{
  "displayName": "NeonStepper",
  "exportName": "default",
  "description": "",
  "tags": {},
  "props": [
    {
      "name": "steps",
      "description": "An ordered list of the step names, these values are used for display purposes.",
      "type": {
        "name": "Array as () => Array<string>"
      },
      "required": true
    },
    {
      "name": "completedIndex",
      "description": "The index of the highest completed step (regardless of the current step the user has selected).\nThis is used to display the completed steps to the user.",
      "type": {
        "name": "number"
      },
      "defaultValue": {
        "func": false,
        "value": "-1"
      }
    },
    {
      "name": "modelValue",
      "description": "The index of the step that is currently selected.",
      "type": {
        "name": "number"
      },
      "defaultValue": {
        "func": false,
        "value": "0"
      }
    }
  ],
  "events": [
    {
      "name": "update:modelValue",
      "description": "Emitted when the selected step is changed.",
      "type": {
        "names": [
          "number"
        ]
      }
    }
  ],
  "sourceFiles": [
    "./src/components/navigation/stepper/NeonStepper.vue",
    "/Users/andrew/dev/projects/neon/src/components/navigation/stepper/NeonStepper.ts"
  ]
}