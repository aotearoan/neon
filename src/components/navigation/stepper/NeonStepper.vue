<template>
  <div :class="`neon-stepper--${color}`" class="neon-stepper">
    <div
      v-for="(step, index) in steps"
      :key="step"
      :class="{
        'neon-stepper__step--has-value': index <= modelValue || index <= completedIndex,
        'neon-stepper__step--active': index === modelValue,
        'neon-stepper__step--display-stepper': index <= completedIndex,
      }"
      class="neon-stepper__step"
    >
      <div
        class="neon-stepper__step-indicator-wrapper"
        @click="index !== modelValue && index <= completedIndex && emit('update:modelValue', index)"
      >
        <div class="neon-stepper__step-indicator">
          <neon-icon v-if="index < completedIndex" class="neon-stepper__step-completed-icon" name="check" />
          <neon-icon v-else-if="index === completedIndex" class="neon-stepper__step-completing-icon" name="dot" />
        </div>
      </div>
      <span
        class="neon-stepper__step-title"
        @click="index !== modelValue && index <= completedIndex && emit('update:modelValue', index)"
      >
        {{ step }}
      </span>
    </div>
  </div>
</template>

<script lang="ts" src="./NeonStepper.ts" />
