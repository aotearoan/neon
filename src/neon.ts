/* components */
export { default as NeonActionMenu } from './components/navigation/action-menu/NeonActionMenu.vue';
export { default as NeonAlert } from './components/feedback/alert/NeonAlert.vue';
export { default as NeonAlertContainer } from './components/feedback/alert/container/NeonAlertContainer.vue';
export { default as NeonAnchor } from './components/navigation/anchor/NeonAnchor.vue';
export { default as NeonBadge } from './components/presentation/badge/NeonBadge.vue';
export { default as NeonButton } from './components/user-input/button/NeonButton.vue';
export { default as NeonCard } from './components/layout/card/NeonCard.vue';
export { default as NeonCardBody } from './components/layout/card/body/NeonCardBody.vue';
export { default as NeonCardFooter } from './components/layout/card/footer/NeonCardFooter.vue';
export { default as NeonCardHeader } from './components/layout/card/header/NeonCardHeader.vue';
export { default as NeonChip } from './components/user-input/chip/NeonChip.vue';
export { default as NeonColor } from './components/user-input/color/NeonColor.vue';
export { default as NeonDialog } from './components/feedback/dialog/NeonDialog.vue';
export { default as NeonDrawer } from './components/layout/drawer/NeonDrawer.vue';
export { default as NeonDropdown } from './components/presentation/dropdown/NeonDropdown.vue';
export { default as NeonDropdownMenu } from './components/navigation/dropdown-menu/NeonDropdownMenu.vue';
export { default as NeonDropZone } from './components/user-input/drop-zone/NeonDropZone.vue';
export {
  default as NeonExpansionIndicator
} from './components/presentation/expansion-indicator/NeonExpansionIndicator.vue';
export { default as NeonExpansionPanel } from './components/presentation/expansion-panel/NeonExpansionPanel.vue';
export { default as NeonFieldGroup } from './components/user-input/field-group/NeonFieldGroup.vue';
export { default as NeonFile } from './components/user-input/file/NeonFile.vue';
export { default as NeonFilterList } from './components/user-input/filter-list/NeonFilterList.vue';
export { default as NeonFooter } from './components/layout/footer/NeonFooter.vue';
export { default as NeonGrid } from './components/layout/grid/NeonGrid.vue';
export { default as NeonGridArea } from './components/layout/grid/grid-area/NeonGridArea.vue';
export { default as NeonIcon } from './components/presentation/icon/NeonIcon.vue';
export { default as NeonInput } from './components/user-input/input/NeonInput.vue';
export { default as NeonInputIndicator } from './components/user-input/input-indicator/NeonInputIndicator.vue';
export { default as NeonLabel } from './components/presentation/label/NeonLabel.vue';
export { default as NeonLinearProgress } from './components/feedback/linear-progress/NeonLinearProgress.vue';
export { default as NeonLink } from './components/navigation/link/NeonLink.vue';
export { default as NeonList } from './components/user-input/list/NeonList.vue';
export { default as NeonLogo } from './components/presentation/logo/NeonLogo.vue';
export { default as NeonMenu } from './components/navigation/menu/NeonMenu.vue';
export { default as NeonModal } from './components/layout/modal/NeonModal.vue';
export { default as NeonNote } from './components/feedback/note/NeonNote.vue';
export {
  default as NeonNotificationCounter
} from './components/feedback/notification-counter/NeonNotificationCounter.vue';
export { default as NeonNumber } from './components/user-input/number/NeonNumber.vue';
export { default as NeonPage } from './components/layout/page/NeonPage.vue';
export { default as NeonPassword } from './components/user-input/password/NeonPassword.vue';
export { default as NeonRangeSlider } from './components/user-input/range-slider/NeonRangeSlider.vue';
export { default as NeonSideNav } from './components/layout/side-nav/NeonSideNav.vue';
export { default as NeonSearch } from './components/user-input/search/NeonSearch.vue';
export { default as NeonSelect } from './components/user-input/select/NeonSelect.vue';
export { default as NeonSkeletonLoader } from './components/feedback/skeleton-loader/NeonSkeletonLoader.vue';
export { default as NeonSlider } from './components/user-input/slider/NeonSlider.vue';
export { default as NeonSplashLoader } from './components/feedback/splash-loader/NeonSplashLoader.vue';
export { default as NeonSwitch } from './components/user-input/switch/NeonSwitch.vue';
export { default as NeonTab } from './components/presentation/tabs/tab/NeonTab.vue';
export { default as NeonTabs } from './components/presentation/tabs/NeonTabs.vue';
export { default as NeonToastContainer } from './components/feedback/alert/container/NeonToastContainer.vue';
export { default as NeonToggle } from './components/user-input/toggle/NeonToggle.vue';
export { default as NeonToggleChip } from './components/user-input/toggle-chip/NeonToggleChip.vue';
export { default as NeonTooltip } from './components/feedback/tooltip/NeonTooltip.vue';
export { default as NeonTopNav } from './components/layout/top-nav/NeonTopNav.vue';
export { default as NeonTreeMenu } from './components/navigation/tree-menu/NeonTreeMenu.vue';

/* enums */
export { NeonAlertLevel } from './common/enums/NeonAlertLevel';
export { NeonAlertPlacement } from './common/enums/NeonAlertPlacement';
export { NeonButtonSize } from './common/enums/NeonButtonSize';
export { NeonButtonStyle } from './common/enums/NeonButtonStyle';
export { NeonChipAction } from './common/enums/NeonChipAction';
export { NeonDropdownPlacement } from './common/enums/NeonDropdownPlacement';
export { NeonDropdownStyle } from './common/enums/NeonDropdownStyle';
export { NeonFunctionalColor } from './common/enums/NeonFunctionalColor';
export { NeonHorizontalPosition } from './common/enums/NeonHorizontalPosition';
export { NeonInputMode } from './common/enums/NeonInputMode';
export { NeonInputType } from './common/enums/NeonInputType';
export { NeonLabelSize } from './common/enums/NeonLabelSize';
export { NeonMode } from './common/enums/NeonMode';
export { NeonOrientation } from './common/enums/NeonOrientation';
export { NeonOutlineStyle } from './common/enums/NeonOutlineStyle';
export { NeonPlacement } from './common/enums/NeonPlacement';
export { NeonPosition } from './common/enums/NeonPosition';
export { NeonResponsive } from './common/enums/NeonResponsive';
export { NeonSize } from './common/enums/NeonSize';
export { NeonSplashLoaderSize } from './common/enums/NeonSplashLoaderSize';
export { NeonState } from './common/enums/NeonState';
export { NeonSwitchStyle } from './common/enums/NeonSwitchStyle';
export { NeonToggleStyle } from './common/enums/NeonToggleStyle';
export { NeonTooltipStyle } from './common/enums/NeonTooltipStyle';
export { NeonVerticalPosition } from './common/enums/NeonVerticalPosition';

/* models */
export type { NeonActionMenuModel } from './common/models/NeonActionMenuModel';
export type { NeonAlertAction } from './common/models/NeonAlertAction';
export type { NeonAlertMessage } from './common/models/NeonAlertMessage';
export type { NeonAlertModel } from './components/feedback/alert/NeonAlertModel';
export type { NeonAvailableSpace } from './common/models/NeonAvailableSpace';
export type { NeonDropdownMenuItem } from './common/models/NeonDropdownMenuItem';
export type { NeonDropdownPlacementObject } from './common/models/NeonDropdownPlacementObject';
export type { NeonFilterListItem } from './common/models/NeonFilterListItem';
export type { NeonFormattedDate } from './common/models/NeonFormattedDate';
export type { NeonGridModel } from './common/models/NeonGridModel';
export type { NeonListItem } from './common/models/NeonListItem';
export type { NeonMenuItem } from './common/models/NeonMenuItem';
export type { NeonMenuModel } from './common/models/NeonMenuModel';
export type { NeonNumberFormatOptions } from './common/models/NeonNumberFormatOptions';
export type { NeonPriorityMenuItem } from './components/navigation/menu/NeonPriorityMenuItem';
export type { NeonSearchOption } from './common/models/NeonSearchOption';
export type { NeonSelectGroup } from './common/models/NeonSelectGroup';
export type { NeonSelectOption } from './common/models/NeonSelectOption';
export type { NeonTabModel } from './common/models/NeonTabModel';
export type { NeonToastMessage } from './common/models/NeonToastMessage';
export type { NeonToastModel } from './components/feedback/alert/NeonToastModel';
export type { NeonToggleModel } from './common/models/NeonToggleModel';
export type { NeonTreeMenuLinkModel, NeonTreeMenuSectionModel } from './common/models/NeonTreeMenuModel';

/* utils */
export { NeonAlertService } from './common/utils/NeonAlertService';
export { NeonClipboardService, NeonClipboardSupport } from './common/utils/NeonClipboardService';
export { NeonClosableUtils } from './common/utils/NeonClosableUtils';
export { NeonDateUtils } from './common/utils/NeonDateUtils';
export { NeonDropdownPlacementUtils } from './common/utils/NeonDropdownPlacementUtils';
export { NeonEventBus } from './common/utils/NeonEventBus';
export { NeonIconRegistry } from './common/utils/NeonIconRegistry';
export { NeonModeUtils } from './common/utils/NeonModeUtils';
export { NeonNumberUtils } from './common/utils/NeonNumberUtils';
export { NeonPlacementUtils } from './common/utils/NeonPlacementUtils';
export { NeonResponsiveUtils } from './common/utils/NeonResponsiveUtils';
export { NeonScrollUtils } from './common/utils/NeonScrollUtils';
export { NeonToastService } from './common/utils/NeonToastService';
export { NeonTooltipPlacementUtils } from './common/utils/NeonTooltipPlacementUtils';

/* register default provided icons */
export { RegisterIcons } from './common/utils/RegisterIcons';
