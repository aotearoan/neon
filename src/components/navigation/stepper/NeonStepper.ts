import { defineComponent } from 'vue';
import NeonIcon from '@/components/presentation/icon/NeonIcon.vue';
import { NeonFunctionalColor } from '@/neon';

/**
 * A stepper component for assisting users in navigating through complex flows.
 */
export default defineComponent({
  name: 'NeonStepper',
  components: { NeonIcon },
  props: {
    /**
     * An ordered list of the step names, these values are used for display purposes.
     */
    steps: { type: Array as () => Array<string>, required: true },
    /**
     * The index of the highest completed step (regardless of the current step the user has selected).
     * This is used to display the completed steps to the user.
     */
    completedIndex: { type: Number, default: -1 },
    /**
     * The index of the step that is currently selected.
     */
    modelValue: { type: Number, default: 0 },
    /**
     * The color of the Stepper component.
     */
    color: { type: String as () => NeonFunctionalColor, default: NeonFunctionalColor.Brand },
  },
  emits: [
    /**
     * Emitted when the selected step is changed.
     * @type {number} The index of the selected step.
     */
    'update:modelValue',
  ],
  setup(_props, { emit }) {
    return {
      emit,
    };
  },
});
