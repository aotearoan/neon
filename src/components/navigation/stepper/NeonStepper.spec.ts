import type { RenderResult } from '@testing-library/vue';
import { fireEvent, render } from '@testing-library/vue';
import NeonStepper from './NeonStepper.vue';
import { NeonFunctionalColor } from '@/neon';

describe('NeonStepper.vue', () => {
  let harness: RenderResult;

  beforeEach(() => {
    harness = render(NeonStepper, {
      props: {
        steps: ['Parties', 'Artwork', 'Additional Terms', 'Duration', 'Review'],
        completedIndex: 3,
        modelValue: 2,
      },
    });
  });

  test('matches snapshot', () => {
    const { html } = harness;
    expect(html()).toMatchSnapshot();
  });

  test('emits input when clicking on previous step', async () => {
    const { emitted, getByText } = harness;

    await fireEvent.click(getByText('Artwork'));
    expect(emitted()['update:modelValue']).toEqual([[1]]);
  });

  test('renders in default color', () => {
    const { container } = harness;
    expect(container.querySelector('.neon-stepper--brand')).toBeDefined();
  });

  test('renders alternative color', async () => {
    const { container, rerender } = harness;
    await rerender({ color: NeonFunctionalColor.Error });
    expect(container.querySelector('.neon-stepper--error')).toBeDefined();
  });

  test('does not emit input when clicking on active step', async () => {
    const { emitted, getByText } = harness;

    await fireEvent.click(getByText('Additional Terms'));
    expect(emitted()['update:modelValue']).toBeUndefined();
  });

  test('does not emit input when clicking on future step', async () => {
    const { emitted, getByText } = harness;

    await fireEvent.click(getByText('Review'));
    expect(emitted()['update:modelValue']).toBeUndefined();
  });
});
