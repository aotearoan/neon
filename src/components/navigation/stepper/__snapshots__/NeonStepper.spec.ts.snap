// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NeonStepper.vue matches snapshot 1`] = `
<div class="neon-stepper--brand neon-stepper">
  <div class="neon-stepper__step--has-value neon-stepper__step--display-stepper neon-stepper__step">
    <div class="neon-stepper__step-indicator-wrapper">
      <div class="neon-stepper__step-indicator">
        <div class="neon-icon--name-check neon-icon neon-stepper__step-completed-icon neon-stepper__step-completed-icon"><svg width="100%" height="100%" viewBox="0 0 24 24" style="fill-rule:evenodd;clip-rule:evenodd;stroke-linejoin:round;stroke-miterlimit:2;">
            <path class="neon-svg--fill" d="M22.596,6.529c0.26,-0.278 0.404,-0.645 0.404,-1.025c0,-0.825 -0.679,-1.504 -1.505,-1.504c-0.381,0 -0.748,0.144 -1.026,0.404l-11.981,11.971l-4.957,-4.953c-0.278,-0.259 -0.645,-0.404 -1.026,-0.404c-0.826,0 -1.505,0.679 -1.505,1.504c-0,0.381 0.144,0.747 0.404,1.025l6.02,6.016c0.584,0.583 1.544,0.583 2.128,0l13.044,-13.034Z" style="fill-rule:nonzero;"></path>
          </svg>
        </div>
      </div>
    </div><span class="neon-stepper__step-title">Parties</span>
  </div>
  <div class="neon-stepper__step--has-value neon-stepper__step--display-stepper neon-stepper__step">
    <div class="neon-stepper__step-indicator-wrapper">
      <div class="neon-stepper__step-indicator">
        <div class="neon-icon--name-check neon-icon neon-stepper__step-completed-icon neon-stepper__step-completed-icon"><svg width="100%" height="100%" viewBox="0 0 24 24" style="fill-rule:evenodd;clip-rule:evenodd;stroke-linejoin:round;stroke-miterlimit:2;">
            <path class="neon-svg--fill" d="M22.596,6.529c0.26,-0.278 0.404,-0.645 0.404,-1.025c0,-0.825 -0.679,-1.504 -1.505,-1.504c-0.381,0 -0.748,0.144 -1.026,0.404l-11.981,11.971l-4.957,-4.953c-0.278,-0.259 -0.645,-0.404 -1.026,-0.404c-0.826,0 -1.505,0.679 -1.505,1.504c-0,0.381 0.144,0.747 0.404,1.025l6.02,6.016c0.584,0.583 1.544,0.583 2.128,0l13.044,-13.034Z" style="fill-rule:nonzero;"></path>
          </svg>
        </div>
      </div>
    </div><span class="neon-stepper__step-title">Artwork</span>
  </div>
  <div class="neon-stepper__step--has-value neon-stepper__step--active neon-stepper__step--display-stepper neon-stepper__step">
    <div class="neon-stepper__step-indicator-wrapper">
      <div class="neon-stepper__step-indicator">
        <div class="neon-icon--name-check neon-icon neon-stepper__step-completed-icon neon-stepper__step-completed-icon"><svg width="100%" height="100%" viewBox="0 0 24 24" style="fill-rule:evenodd;clip-rule:evenodd;stroke-linejoin:round;stroke-miterlimit:2;">
            <path class="neon-svg--fill" d="M22.596,6.529c0.26,-0.278 0.404,-0.645 0.404,-1.025c0,-0.825 -0.679,-1.504 -1.505,-1.504c-0.381,0 -0.748,0.144 -1.026,0.404l-11.981,11.971l-4.957,-4.953c-0.278,-0.259 -0.645,-0.404 -1.026,-0.404c-0.826,0 -1.505,0.679 -1.505,1.504c-0,0.381 0.144,0.747 0.404,1.025l6.02,6.016c0.584,0.583 1.544,0.583 2.128,0l13.044,-13.034Z" style="fill-rule:nonzero;"></path>
          </svg>
        </div>
      </div>
    </div><span class="neon-stepper__step-title">Additional Terms</span>
  </div>
  <div class="neon-stepper__step--has-value neon-stepper__step--display-stepper neon-stepper__step">
    <div class="neon-stepper__step-indicator-wrapper">
      <div class="neon-stepper__step-indicator">
        <div class="neon-icon--name-dot neon-icon neon-stepper__step-completing-icon neon-stepper__step-completing-icon"><svg width="32" height="32" viewBox="0 0 32 32" xmlns="http://www.w3.org/2000/svg">
            <circle cx="16" cy="16" r="15.5" class="neon-svg--fill neon-svg--stroke"></circle>
          </svg>
        </div>
      </div>
    </div><span class="neon-stepper__step-title">Duration</span>
  </div>
  <div class="neon-stepper__step">
    <div class="neon-stepper__step-indicator-wrapper">
      <div class="neon-stepper__step-indicator">
        <!--v-if-->
      </div>
    </div><span class="neon-stepper__step-title">Review</span>
  </div>
</div>
`;
